### PR TITLE
[Indexing] Add Indexing dialect

### PR DESCRIPTION
**Pull request by makslevental**
Add boilerplate for `Indexing` dialect.

### CODE REVIEW

**Review by nicolasvasilache — APPROVED** [review]

File: include/structured/Dialect/Indexing/IR/IndexingDialect.td [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+include "mlir/IR/OpBase.td"
+
+//===----------------------------------------------------------------------===//
+// Dialect definition
+//===----------------------------------------------------------------------===//
+
+def Indexing_Dialect
+: Dialect {
+  let name = "indexing";
+  let cppNamespace = "::mlir::structured";
```
Comment (ingomueller-net) [inline include/structured/Dialect/Indexing/IR/IndexingDialect.td:21]: I suggest to do the following:

```suggestion
  let cppNamespace = "::mlir::indexing";
```

File: include/structured/Dialect/Indexing/IR/IndexingDialect.td [hunk 2]
```diff
@@ -0,0 +1,25 @@
+//===-- IndexingDialect.td - Indexing dialect ------------*- tablegen -*-===//
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef INDEXING_DIALECT_INDEXING_IR_INDEXINGDIALECT
```
Comment (ingomueller-net) [inline include/structured/Dialect/Indexing/IR/IndexingDialect.td:9]: These need to be adapted to the new folder layout.

```suggestion
#ifndef STRUCTURED_DIALECT_INDEXING_IR_INDEXINGDIALECT
```

File: lib/Dialect/Indexing/IR/Indexing.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+// Indexing dialect
+//===----------------------------------------------------------------------===//
+
+#include "structured/Dialect/Indexing/IR/IndexingOpsDialect.cpp.inc"
+
+void IndexingDialect::initialize() {
+#define GET_OP_LIST
+  addOperations<
+
+#include "structured/Dialect/Indexing/IR/IndexingOps.cpp.inc"
```
Comment (ingomueller-net) [inline lib/Dialect/Indexing/IR/Indexing.cpp:32]: I think people usually don't leave empty lines but I only care for consistency purposes...

**Review by ingomueller-net — COMMENTED** [review]
Looks good as far as I can tell: I don't have a strong enough sense about the whole problem space to say something about the op but the boilerplate code etc looks good!

**Comment by makslevental** [comment]
> Looks good as far as I can tell: I don't have a strong enough sense about the whole problem space to say something about the op but the boilerplate code etc looks good!

Just a btw - that op is just upstream [`tensor.gather`](https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorgather-mlirtensorgatherop) (added by @nicolasvasilache) but with the small addition of `DeclareOpInterfaceMethods<InferTypeOpInterface>`. I plan to upstream (very shortly) whatever gets worked out here regarding its semantics.

File: examples/advanced-indexing.ipynb [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+    "assert np.array_equal(abcd[:, *(1, 0), :], c)\n",
+    "assert np.array_equal(abcd[:, *(1, 1), :], d)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "61ef3393",
+   "metadata": {},
+   "source": [
+    "It's also important to understand what `abcd` is; while the original array has dimensions $2 \\times 30 \\times 4$ while `abcd` has distinct (different from `input`) interior dimensions. $2 \\times 2$ seems somehow smaller/related to $30$ but it's not; the indexing dimensions determine output side _at those dimensions_:"
```
Comment (ingomueller-net) [inline examples/advanced-indexing.ipynb:130]: Again: the dimension given in the text doesn't seem to fit that used in the code (2 vs 3).

**Review by ingomueller-net — COMMENTED** [review]
@makslevental: I must have overseen this file when I first reviewed this PR, probably because it was collapsed in the diff. Here are some minor comments to the first section.

This was super useful to go through for me, though I am still struggling to produce intuitive representations of these multi-dimensional slices in my head...

**Review by ingomueller-net — COMMENTED** [review]
@makslevental: I must have overseen this file when I first reviewed this PR, probably because it was collapsed in the diff. Here are some minor comments to the first section.

This was super useful to go through for me, though I am still struggling to produce intuitive representations of these multi-dimensional slices in my head...

**Comment by makslevental** [comment]
> This was super useful to go through for me, though I am still struggling to produce intuitive representations of these multi-dimensional slices in my head...

I don't think it's useful to in fact picture the slices (unless you generally can picture things in n-D lol). Instead I think it's better to just focus on the mapping between the numpy syntax and the indexing dialect syntax. So I made this explainer for `indexing.gather` (I will add it to the repo when it comes time for docs):

numpy_adv_vs_gather

**Pull request merged**